src/Form/ConstraintForm.php

```diff
@@ -0,0 +1,233 @@
+<?php
+
+namespace Drupal\password_enhancements\Form;
+
+use Drupal\Component\Plugin\Exception\PluginException;
+use Drupal\Core\Config\Entity\ConfigEntityStorageInterface;
+use Drupal\Core\Entity\EntityForm;
+use Drupal\Core\Form\FormStateInterface;
+use Drupal\Core\Messenger\MessengerInterface;
+use Drupal\password_enhancements\PasswordConstraintPluginManager;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * Password constraint config entity form.
+ */
+class ConstraintForm extends EntityForm {
+
+  /**
+   * Config entity storage.
+   *
+   * @var \Drupal\Core\Config\Entity\ConfigEntityStorageInterface
+   */
+  protected $constraintConfigEntityStorage;
+
+  /**
+   * Password constraint plugin manager.
+   *
+   * @var \Drupal\password_enhancements\PasswordConstraintPluginManager
+   */
+  protected $passwordConstraintPluginManager;
+
+  /**
+   * Constructs the entity form for password_enhancements_constraint entity.
+   *
+   * @param \Drupal\Core\Config\Entity\ConfigEntityStorageInterface $constraint_config_entity_storage
+   *   Config entity storage.
+   * @param \Drupal\password_enhancements\PasswordConstraintPluginManager $password_constraint_plugin_manager
+   *   Password constraint plugin manager.
+   * @param \Drupal\Core\Messenger\MessengerInterface $messenger
+   *   Messenger.
+   */
+  public function __construct(ConfigEntityStorageInterface $constraint_config_entity_storage, PasswordConstraintPluginManager $password_constraint_plugin_manager, MessengerInterface $messenger) {
+    $this->constraintConfigEntityStorage = $constraint_config_entity_storage;
+    $this->passwordConstraintPluginManager = $password_constraint_plugin_manager;
+    $this->messenger = $messenger;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public static function create(ContainerInterface $container): ConstraintForm {
+    return new static(
+      $container->get('entity_type.manager')->getStorage('password_enhancements_constraint'),
+      $container->get('plugin.manager.password_constraint'),
+      $container->get('messenger')
+    );
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function form(array $form, FormStateInterface $form_state): array {
+    $form = parent::form($form, $form_state);
+
+    $form['#title'] = $this->t('Edit constraint');
+
+    $options = [];
+    foreach ($this->passwordConstraintPluginManager->getDefinitions() as $definition) {
+      $options[$definition['id']] = $definition['name'];
+    }
+
+    $form['policy'] = [
+      '#type' => 'hidden',
+      '#value' => $this->getRequest()->get('password_enhancements_policy')->id(),
```

Comment:
DI the request_stack service, otherwise $this->getRequest() silently calls the global static container to retrieve it.